### PR TITLE
Fix PeriodIndex._shallow_copy allowing object-dtype

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/core/indexes/period.py [hunk 1]
```diff
@@ -272,19 +272,13 @@ def _shallow_copy(self, values=None, **kwargs):
...
                 values = PeriodArray(values, freq=self.freq)
             else:
-                # in particular, I would like to avoid period_array here.
-                # Some people seem to be calling use with unexpected types
-                # Index.difference -> ndarray[Period]
-                # DatetimelikeIndexOpsMixin.repeat -> ndarray[ordinal]
-                # I think that once all of Datetime* are EAs, we can simplify
-                # this quite a bit.
-                values = period_array(values, freq=self.freq)
+                raise TypeError(type(values), getattr(values, "dtype", None))
```
Comment (jreback) [inline pandas/core/indexes/period.py]: what triggers this?

Comment (jbrockmendel) [reply pandas/core/indexes/period.py]: After this PR, this is never reached.  In master we get there with ndarray[object] via dropna and symmetric_difference. The PeriodIndex set ops PRs over the last couple days have been aimed at this

Comment (jreback) [reply pandas/core/indexes/period.py]: ok can u make a comment about this then 

should this just be an assertion ?

**Comment by jreback** [comment]
thanks

**Pull request merged**